src/uct/tcp/tcp_sockcm_ep.c

```diff
@@ -709,6 +709,59 @@ static ucs_status_t uct_tcp_sockcm_ep_client_init(uct_tcp_sockcm_ep_t *cep,
     return status;
 }
 
+static ucs_status_t uct_tcp_sockcm_ep_server_update_cm(uct_tcp_sockcm_ep_t *tcp_ep,
+                                                       uct_cm_h params_cm)
+{
+    uct_tcp_sockcm_t *tcp_sockcm        = uct_tcp_sockcm_ep_get_cm(tcp_ep);
+    uct_tcp_sockcm_t *params_tcp_sockcm = ucs_derived_of(params_cm, uct_tcp_sockcm_t);
+    ucs_async_context_t *async_ctx;
+    ucs_status_t status;
+
+    /* check if the server opened this ep, to the client, on a CM that is
+     * different from the one it created its internal ep on earlier, when it
+     * received the connection request from the client (the cm used by its listener) */
+    if (tcp_sockcm == params_tcp_sockcm) {
+        return UCS_OK;
+    }
+
+    /* set the server's ep to use the new cm and its iface */
+    tcp_ep->super.super.super.iface = &params_cm->iface.super;
+
+    status = uct_base_ep_stats_switch_iface(&tcp_ep->super.super, &params_cm->iface);
+    if (status != UCS_OK) {
+        ucs_error("failed to switch iface (%p) on ep %p : %s",
+                  &params_cm->iface, tcp_ep, ucs_status_string(status));
+        goto out;
+    }
```

Comment:
need to allocate ep stats only after ep is created on server side (and not while the cep is still treated as conn_request)

Reply:
it was already allocated when the tcp_sockcm's ep ctor was called:
https://github.com/openucx/ucx/blob/master/src/uct/tcp/tcp_sockcm_ep.c#L716     -->
https://github.com/openucx/ucx/blob/master/src/uct/base/uct_iface.c#L582

Reply:
it shouldn't be allocated at the stage where cep is only a conn_request

Reply:
but isn't it always part of the ep creation? i mean, it this place, stats were already allocated. i'm only updating the iface for it

Reply:
from API perspective, the ep is still not created

Reply:
so it wasnt ok even before the change in this pr?

Reply:
yes

Reply:
then shouldn't stats' allocation be removed from uct_base_ep_t ctor (regardless of this pr)? 
it's always created during uct_ep_create while \'the ep is still not created\'

Reply:
this will affect other transports
can we avoid calling ep base ctor for server side when creating conn_request?

Reply:
>this will affect other transports

yes, this is what i meant. but then isn't their behavior wrong too? they do the same...

>can we avoid calling ep base ctor for server side when creating conn_request?

i'm not calling ep_base ctor on the uct_ep_create call here.
when the server first creates its 'internal' ep, it calls the ep_base ctor.
then during uct_ep_create, here, if need to reset the cm, then i "reset" the cm for everything it was used in - stats and cm->iface.

Reply:
@yosefe wdyt?

Reply:
"when the server first creates its 'internal' ep, it calls the ep_base ctor." - need to avoid this

Reply:
and so it should call the ep_base ctor in uct_ep_create?

Reply:
yes, only in uct_ep_create

Reply:
in uct_ep_create, the ctor isnt called for the server side so i dont have 'self' and cant call the ep_base ctor. i can invoke the same functions it does instead